testutil/network/gatewaynet/network.go

```diff
@@ -0,0 +1,190 @@
+	require.NotNil(t, resp)
+	require.NotNil(t, resp.TxHash)
+	require.Equal(t, uint32(0), resp.Code)
+}
+
+// UndelegateAppFromGateway delegates the application by address from the gateway by address.
+func (memnet *inMemoryNetworkWithGateways) UndelegateAppFromGateway(
+	t *testing.T,
+	appBech32 string,
+	gatewayBech32 string,
+) {
+
+	args := []string{
+		gatewayBech32,
+		fmt.Sprintf("--%s=%s", flags.FlagFrom, appBech32),
+		fmt.Sprintf("--%s=true", flags.FlagSkipConfirmation),
+		fmt.Sprintf("--%s=%s", flags.FlagBroadcastMode, flags.BroadcastSync),
+		fmt.Sprintf("--%s=%s", flags.FlagFees, sdk.NewCoins(sdk.NewCoin(memnet.GetCosmosNetworkConfig(t).BondDenom, math.NewInt(10))).String()),
+	}
+	responseRaw, err := testcli.ExecTestCLICmd(memnet.GetClientCtx(t), cli.CmdUndelegateFromGateway(), args)
+	require.NoError(t, err)
+	var resp sdk.TxResponse
+	require.NoError(t, memnet.GetCosmosNetworkConfig(t).Codec.UnmarshalJSON(responseRaw.Bytes(), &resp))
+	require.NotNil(t, resp)
+	require.NotNil(t, resp.TxHash)
+	require.Equal(t, uint32(0), resp.Code)
+}
+
+// Start initializes the in-memory network and performs the following setup:
+//   - populates a new in-memory keyring with a sufficient number of pre-generated accounts.
+//   - configures the application module's genesis state using addresses corresponding
+//     to config.NumApplications number of the same pre-generated accounts which were
+//     added to the keyring.
+//   - configures the supplier module's genesis state using addresses corresponding to
+//     config.NumSuppliers number of the same pre-generated accounts which were added
+//     to the keyring.
+//   - creates the on-chain accounts in the accounts module which correspond to the
+//     pre-generated accounts which were added to the keyring.
+func (memnet *inMemoryNetworkWithGateways) Start(_ context.Context, t *testing.T) {
+	t.Helper()
+
+	// Application module genesis state fixture data generation is independent
+	// of that of the supplier module.
+	if memnet.Config.AppSupplierPairingRatio > 0 {
+		panic("AppSupplierPairingRatio must be 0 for inMemoryNetworkWithGateways, use NumApplications instead")
+	}
+
+	memnet.InitializeDefaults(t)
+	memnet.CreateKeyringAccounts(t)
+
+	// Configure gateway and application module genesis states.
+	memnet.configureGatewayModuleGenesisState(t)
+	memnet.configureAppModuleGenesisState(t)
+
+	memnet.Network = network.New(t, *memnet.GetCosmosNetworkConfig(t))
+
+	memnet.CreateOnChainAccounts(t)
+}
+
+// TODO_IN_THIS_COMMIT: fix comment...
+// GatewayModuleGenesisStateWithAddresses generates a GenesisState object with
+// a gateway list full of gateways with the given addresses.
+func (memnet *inMemoryNetworkWithGateways) configureGatewayModuleGenesisState(t *testing.T) {
+	t.Helper()
+
+	gatewayGenesisState := gatewaytypes.DefaultGenesis()
+	for gatewayIdx := 0; gatewayIdx < memnet.Config.NumGateways; gatewayIdx++ {
+		stake := sdk.NewCoin("upokt", sdk.NewInt(int64(gatewayIdx)))
```

Comment:
```suggestion
		stake := sdk.NewCoin("upokt", sdk.NewInt(int64(gatewayIdx)+1))
```
Zero stake results in an error iirc.

Might be better to take this approach

```go
stake := &sdk.Coin{Denom: "upokt", Amount: sdk.NewInt(10000)}
```